dartagnan/src/main/java/com/dat3m/dartagnan/wmm/Wmm.java

```diff
@@ -260,71 +209,115 @@ private void simplifyAssociatives(Class<? extends Definition> cls, BiFunction<Re
+            case FR -> composition(r, getOrCreatePredefinedRelation(RFINV), getOrCreatePredefinedRelation(CO));
+            case MM -> product(r, Tag.MEMORY, Tag.MEMORY);
+            case MV -> product(r, Tag.MEMORY, Tag.VISIBLE);
+            case IDDTRANS -> new TransitiveClosure(r, getOrCreatePredefinedRelation(IDD));
+            case DATA -> intersection(r, getOrCreatePredefinedRelation(IDDTRANS), getOrCreatePredefinedRelation(MM));
             case ADDR -> {
-                Relation addrdirect = getRelation(ADDRDIRECT);
-                Relation comp = addDefinition(composition(newRelation(), getRelation(IDDTRANS), addrdirect));
+                Relation addrdirect = getOrCreatePredefinedRelation(ADDRDIRECT);
+                Relation comp = addDefinition(composition(newRelation(), getOrCreatePredefinedRelation(IDDTRANS), addrdirect));
                 Relation union = addDefinition(union(newRelation(), addrdirect, comp));
-                yield intersection(r, union, getRelation(MM));
+                yield intersection(r, union, getOrCreatePredefinedRelation(MM));
             }
             case CTRL -> {
-                Relation comp = addDefinition(composition(newRelation(), getRelation(IDDTRANS), getRelation(CTRLDIRECT)));
-                yield intersection(r, comp, getRelation(MV));
+                Relation comp = addDefinition(composition(newRelation(), getOrCreatePredefinedRelation(IDDTRANS),
+                        getOrCreatePredefinedRelation(CTRLDIRECT)));
+                yield intersection(r, comp, getOrCreatePredefinedRelation(MV));
             }
-            case POLOC -> intersection(r, getRelation(PO), getRelation(LOC));
-            case RFE ->  intersection(r, getRelation(RF), getRelation(EXT));
-            case RFI -> intersection(r, getRelation(RF), getRelation(INT));
-            case COE -> intersection(r, getRelation(CO), getRelation(EXT));
-            case COI -> intersection(r, getRelation(CO), getRelation(INT));
-            case FRE -> intersection(r, getRelation(FR), getRelation(EXT));
-            case FRI -> intersection(r, getRelation(FR), getRelation(INT));
+            case POLOC -> intersection(r, getOrCreatePredefinedRelation(PO), getOrCreatePredefinedRelation(LOC));
+            case RFE ->  intersection(r, getOrCreatePredefinedRelation(RF), getOrCreatePredefinedRelation(EXT));
+            case RFI -> intersection(r, getOrCreatePredefinedRelation(RF), getOrCreatePredefinedRelation(INT));
+            case COE -> intersection(r, getOrCreatePredefinedRelation(CO), getOrCreatePredefinedRelation(EXT));
+            case COI -> intersection(r, getOrCreatePredefinedRelation(CO), getOrCreatePredefinedRelation(INT));
+            case FRE -> intersection(r, getOrCreatePredefinedRelation(FR), getOrCreatePredefinedRelation(EXT));
+            case FRI -> intersection(r, getOrCreatePredefinedRelation(FR), getOrCreatePredefinedRelation(INT));
```

Comment:
I start to believe that it might be better to get rid of all these. I would be better to have them defined directly in cat (maybe via a proper `include X.cat`)

Reply:
Definitely, but that requires adding support for includes :).

Reply:
Not really ... we could directly define this directly in each cat file. The question is if ww want to do that in favor of simplifying the code, or we postpone it until we have support for includes.

Reply:
I think you were always in favour of being able to drop in cat-files from herd7, but without includes this would not work anymore. If you are willing to rewrite all cat-models, go for it :P.

Reply:
Fair point. For compatibility with herd, let's keep the code as it is until we support includes